### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by tudor-alexa99**
Request: You are in charge of creating a software for a Car Cleaning Service. It takes 2 hours to process one request and can process two cars at a time. Your software needs to let the users know when to pick up the car. 

Write a ruby program which determines the pickup date and time, knowing the functioning schedule of the Car Cleaning Service.

Please note that there can be edge cases when a customer may need to pick up the car the next day or even next week.
Also, you need to write unit tests to cover all the scenarios for the car schedule.

### CODE REVIEW

File: .idea/CarService.iml [hunk 1]
```diff
@@ -11,5 +11,6 @@
     </content>
```
Comment (VictorMotogna) [inline .idea/CarService.iml:11]: usually, these files need to be ignored. you can add whole `.idea/` folder to `.gitignore`

File: service.rb [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+          Time.new(
+              date.year,
+              date.month,
+              date.day,
+              CLOSE_AT_WEEKDAY.split(":")[0],
+              CLOSE_AT_WEEKDAY.split(":")[1])
+    end
+  end
+
+  def user_brings_car()
```
Comment (florinionce) [inline service.rb]: From ruby style guide: 
> Use def with parentheses when there are parameters. Omit the parentheses when the method doesn’t accept any parameters.

Read more here  => https://github.com/rubocop-hq/ruby-style-guide#method-parens

File: service.rb [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+  def initialize
+    # check the current day and set the closing time regarding week time and weekend hours
+    @starting_hours, @ending_hours = get_daily_working_hours
+
+    @worker_1_available = @starting_hours.dup # make a dup
+    @worker_2_available = @starting_hours.dup
+  end
+
+  def get_daily_working_hours(date = Time.now)
```
Comment (florinionce) [inline service.rb]: is there a reason why this method is public?

File: service.rb [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    # get the closing hours for that day
+    closing_hours = get_daily_working_hours(process_car_time)[1]
+
+    # if the car cannot be processed within the day, move it to the next available day
+    if closing_hours < process_car_time
+      extra_time = process_car_time - closing_hours
+      return process_next_day(extra_time, process_car_time)
+    else
+      # I know the returns are redundant, I added them to emphasise that the value will be used outside
+      return process_today(process_car_time)
```
Comment (florinionce) [inline service.rb]: Well, it's not consistent to use the return, you used it for some methods and didn't use them for the others

File: service.rb [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    if next_day[0].strftime("%A") == "Sunday"
+      next_day = get_daily_working_hours(process_car_time + 2 * 24 * 3600)
+    end
+
+    available_worker = next_day[0] + extra_time
+    # [@worker_1_available, @worker_2_available].min = available_worker <== for some reason, this does not work
+    set_available_worker(available_worker)
+
+    # the pick-up time will be this one
+    return [@worker_1_available, @worker_2_available].max
```
Comment (florinionce) [inline service.rb]: here, for example, you don't need to emphasize the return value since it's obvious

File: spec/car_service_spec.rb [hunk 6]
```diff
@@ -0,0 +1,72 @@
+require_relative '../service'
+
+
+describe Service do
+  context "During week time " do
```
Comment (florinionce) [inline spec/car_service_spec.rb:4]: you have an extra space at the end of the file

File: spec/car_service_spec.rb [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+      @service = Service.new
+    end
+
+    it 'When two clients come at the same time' do
+      expect(@service.add_reservation).to eq Time.now + 2 * 3600
+      expect(@service.add_reservation).to eq Time.now + 2 * 3600
+    end
+
+    it 'When a client comes an hour before closing time ' do
```
Comment (florinionce) [inline spec/car_service_spec.rb:17]: another space

File: service.rb [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+              date.year,
+              date.month,
+              date.day,
+              CLOSE_AT_WEEKDAY.split(":")[0],
+              CLOSE_AT_WEEKDAY.split(":")[1])
+    end
+  end
+
+  def user_brings_car()
+    # Method to be used as a "pretty print". Prints the result of the add_reservation to the user
```
Comment (VictorMotogna) [inline service.rb]: I like the idea for this method, GJ

File: service.rb [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+          date.day,
+          OPEN_FROM_WEEKDAY.split(":")[0],
+          OPEN_FROM_WEEKDAY.split(":")[1]),
+
+          Time.new(
+              date.year,
+              date.month,
+              date.day,
+              CLOSE_AT_WEEKDAY.split(":")[0],
+              CLOSE_AT_WEEKDAY.split(":")[1])
```
Comment (VictorMotogna) [inline service.rb]: there is a lot of duplicate code in this `case when`. You could've used the `case when` to define 2 variables, like this:
```ruby
case week_day
when "Saturday"
  starting_hours = OPEN_FROM_WEEKEND.split(":")[0]
  starting_mins = OPEN_FROM_WEEKEND.split(":")[1]
  # closing_hours = ..
  # closing_mins = ..
when "Sunday
  starting_hours = 0
  closing_hours = 0
  starting_mins = 0
  closing_mins = 0
else
  starting_hours = OPEN_FROM_WEEKDAY.split(":")[0]
  starting_mins = OPEN_FROM_WEEKDAY.split(":")[1]
  # same for closing ..
```

and then return 
```ruby
Time.new(
  date.year,
  date.month,
  date.day,
  starting_hours,
  starting_mins,
  Time.new(
    date.year,
    date.month,
    date.day,
    closing_hours,
    closing_mins
)
```

File: spec/car_service_spec.rb [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+  context "During week time " do
+    before do
+      @monday = Time.new(2020, 6, 1, 10, 00, 00)
+      allow(Time).to receive(:now).and_return(@monday)
+
+      @service = Service.new
+    end
+
+    it 'When two clients come at the same time' do
+      expect(@service.add_reservation).to eq Time.now + 2 * 3600
```
Comment (florinionce) [inline spec/car_service_spec.rb]: I would say that it's not very intuitive the way this looks. If I  see this method name `add_reservation`, I don't expect it to return the pickup time. Maybe it's best if this method is rather setting a `pickup_time` object or something and then you would get it from there. Or simply rename the method

File: service.rb (hunk 8 above)
Comment (tudor-alexa99) [reply service.rb]: Thanks!

File: spec/car_service_spec.rb (hunk 10 above)
Comment (tudor-alexa99) [reply spec/car_service_spec.rb]: I did rename the method
